File: jsyd_102/src/SSPInput.java
import java.io.*;
import java.util.*;

public class SSPInput extends SSModule implements SydConstants
{
  public int           pNbr = 1;
  public String        desc = "input value";
  public ExpRec        defExp = InitExp("");

  // dynamic
  double        pValue;

  public SSPInput(ModList mList)
  {
    super(MT_PInput, mList);
    DescribeLink(0, "INVALID", "???");
  }

  void  Copy(SSModule ss)
  {
    SSPInput sa = (SSPInput) ss;
    super.Copy(ss);
    CopyExp(sa.defExp, defExp);
    pNbr = sa.pNbr;
    desc = sa.desc;
  }

  void Save(PrintWriter ar) throws IOException
  {
    super.Save(ar);
    ar.println("PIDEF " + defExp.exp);
    ar.println("PINN " + pNbr);
    ar.println("PIND " + desc);
  }

  void Load(BufferedReader ar) throws IOException
  {
    defExp = LoadExp(ar,"PIDEF");
    pNbr = Integer.parseInt(getFirstToken(ar,"PINN"));
    desc = LoadString(ar,"PIND");
  }

  double GenerateOutput(SSModule callingMod)
  {
    return pValue;
  }

  void Reset(SSModule callingMod)
  {
    super.Reset(callingMod);
    ResetExp(defExp, callingMod);
    if (callingMod.IsInScore()) {
      pValue = callingMod.GetInstParameter(pNbr);
    }
    else {
      pValue = SolveExp(defExp, callingMod);
    }
    lastRightSample = pValue;
  }

}

File: csyd_102/src/sspinput.cpp
// Folder Input

// sspinput.cp

#include "ss.h"
#include "sspinput.h"
// #include "MainWin.h"

SSPInput::SSPInput(ModList * mList, short h, short v) : SSModule(MT_PInput, mList, h, v)
{
	pNbr = 1;
	ClearExp(&defExp);
	strcpy(desc,"input value");
	DescribeLink(0, "INVALID", "???",0xFFFF,0xFFFF,0xFFFF);
}

void	SSPInput::Copy(SSModule *ss)
{
	SSPInput *sa = (SSPInput *) ss;
	SSModule::Copy(ss);
	CopyExp(&sa->defExp, &defExp);
	pNbr = sa->pNbr;
	strcpy(desc,sa->desc);
}


/*
void SSPInput::Save(FILE* ar)
{
	SSModule::Save(ar);
	WriteFileLine(ar,"PIDEF %s\r",defExp.exp);
	WriteFileLine(ar,"PINN %d\r",pNbr);
	WriteFileLine(ar,"PIND %s\r",desc);
}
*/

void SSPInput::Load(FILE* ar)
{
	char	*p, tbuf[256];
	LoadExp(ar,"PIDEF", &defExp);

	p = parList->GetNextInputLine(ar,"PINN", tbuf);
	sscanf(p,"PINN %d",&pNbr);

	p = parList->GetNextInputLine(ar,"PIND", tbuf);
	sscanf(p,"PIND %s",desc);
}

ssfloat SSPInput::GenerateOutput(SSModule *callingMod)
{
	return pValue;
}

void SSPInput::Reset(SSModule *callingMod)
{
	SSModule::Reset(callingMod);

	this->callingMod = callingMod;

	ResetExp(&defExp,callingMod);

	if (callingMod->IsInScore()) {
		pValue = callingMod->GetInstParameter(pNbr);
	}
	else {
		pValue = SolveExp(&defExp,callingMod);
	}
  lastRightSample = pValue;
}

#ifdef UI_FEATURES

char*	SSPInput::GetLabel()
{
	if (desc[0])
		return desc;
	else
		return NULL;
}

Boolean SSPInput::GetOverlay(char *overlay, int *x, int *y)
{
	*x = cellBounds.left+11;
	*y = cellBounds.top+20;
	sprintf(overlay,"%d",pNbr);
	return true;
}
#endif

